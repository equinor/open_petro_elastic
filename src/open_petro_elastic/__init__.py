import importlib.metadata
from .float_vectorize import float_vectorize


__version__ = importlib.metadata.version(__package__ or __name__)

__all__ = [
    "float_vectorize",
]
