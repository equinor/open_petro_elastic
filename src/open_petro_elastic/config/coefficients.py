from typing import Optional, Union

try:
    from pydantic.v1.dataclasses import dataclass
except ImportError:
    from pydantic.dataclasses import dataclass

from open_petro_elastic.material import polyval_material

from .pydantic_config import PetroElasticConfig
from .vector_type import Array


@dataclass(config=PetroElasticConfig)
class ModuliCoefficients:
    density: Optional[Array[float]] = None
    bulk_modulus: Optional[Array[float]] = None
    shear_modulus: Optional[Array[float]] = None

    @property
    def as_list(self):
        return [self.density, self.bulk_modulus, self.shear_modulus]

    @property
    def use_moduli(self):
        return True

    def polyfit(self, value, material):
        return polyval_material(
            value,
            material,
            self.density,
            self.bulk_modulus,
            self.shear_modulus,
            use_moduli=True,
        )


@dataclass(config=PetroElasticConfig)
class VelocityCoefficients:
    density: Optional[Array[float]] = None
    secondary_velocity: Optional[Array[float]] = None
    primary_velocity: Optional[Array[float]] = None

    @property
    def as_list(self):
        return [self.density, self.primary_velocity, self.secondary_velocity]

    @property
    def use_moduli(self):
        return False

    def polyfit(self, value, material):
        return polyval_material(
            value,
            material,
            self.density,
            self.primary_velocity,
            self.secondary_velocity,
            use_moduli=False,
        )


@dataclass(config=PetroElasticConfig)
class VpOverVsCoefficients:
    density: Optional[Array[float]] = None
    vp_over_vs: Optional[Array[float]] = None
    bulk_modulus: Optional[Array[float]] = None

    @property
    def as_list(self):
        return [self.density, self.bulk_modulus, self.vp_over_vs]


PolyfitCoefficients = Union[ModuliCoefficients, VelocityCoefficients]
Coefficients = Union[VpOverVsCoefficients, ModuliCoefficients, VelocityCoefficients]
