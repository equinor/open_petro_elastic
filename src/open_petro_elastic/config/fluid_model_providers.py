import collections

import pkg_resources

import open_petro_elastic.material.batzle_wang as default
from open_petro_elastic.material import span_wagner
from open_petro_elastic.material.material import vectorize_material


class BatzleWangFluidModelProvider:
    def oil(self, temperature, pressure, reference_density, gas_oil_ratio, gas_gravity):
        return vectorize_material(default.oil)(
            temperature, pressure, reference_density, gas_oil_ratio, gas_gravity
        )

    def gas(self, temperature, pressure, gas_gravity):
        return vectorize_material(default.gas)(temperature, pressure, gas_gravity)

    def brine(self, temperature, pressure, salinity):
        return vectorize_material(default.brine)(temperature, pressure, salinity)

    def condensate(
        self, temperature, pressure, oil_reference_density, gas_oil_ratio, gas_gravity
    ):
        raise NotImplementedError(
            "Default Fluid model provider has no condensate model."
        )

    def carbon_dioxide(self, *args, **kwargs):
        raise NotImplementedError(
            "Default fluid model provider has no carbon dioxide model"
        )


class SpanWagnerFluidModelProvider:
    def oil(self, *args, **kwargs):
        raise NotImplementedError("Span & Wagner fluid model provider has no oil model")

    def gas(self, *args, **kwargs):
        raise NotImplementedError("Span & Wagner fluid model provider has no gas model")

    def brine(self, *args, **kwargs):
        raise NotImplementedError(
            "Span & Wagner fluid model provider has no brine model"
        )

    def condensate(self, *args, **kwargs):
        raise NotImplementedError(
            "Span & Wagner fluid model provider has no condensate model"
        )

    def carbon_dioxide(self, temperature, pressure, interpolate_density):
        from open_petro_elastic.material.conversions import celsius_to_kelvin

        temperature = celsius_to_kelvin(temperature)
        pressure = pressure / 1e6
        return span_wagner.carbon_dioxide(
            temperature,
            pressure,
            None,
            force_vapor="auto",
            interpolate=interpolate_density,
        )


model_names = [
    entry_point.name
    for entry_point in pkg_resources.iter_entry_points(
        "open_petro_elastic.fluid_model_providers"
    )
]

duplicate_names = [
    item for item, count in collections.Counter(model_names).items() if count > 1
]

if duplicate_names:
    raise ImportError(
        "There is a name collision in open_petro_elastic.fluid_model_providers,"
        "check installed plugins.\n"
        f"The following fluid_model_providers have duplicates: ${duplicate_names}"
    )


fluid_model_providers = {
    entry_point.name: entry_point.load()()
    for entry_point in pkg_resources.iter_entry_points(
        "open_petro_elastic.fluid_model_providers"
    )
}

fluid_model_providers["default"] = fluid_model_providers["batzle_wang"]
