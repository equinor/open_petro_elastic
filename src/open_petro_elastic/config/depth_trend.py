from typing import Optional

import numpy as np

try:
    from pydantic.v1.dataclasses import dataclass
except ImportError:
    from pydantic.dataclasses import dataclass

from typing_extensions import Literal

from .coefficients import PolyfitCoefficients
from .pydantic_config import PetroElasticConfig
from .vector_type import Array


@dataclass(config=PetroElasticConfig)
class DepthTrend:
    """
    Adjusts given dry rock to depth. Difference is
    calculated from given polynomial.
    :param reference_dry_rock: Dry rock to adjust at reference depth.
    :param depth: Difference in depth from reference.
    :param coefficients: The polynomial coefficients to used to adjust the material,
    such that the terms in the polynomial are p[i,j] * depth**i *
    mineral.bulk_modulus ** j.
    """

    depth: Array[float]
    coefficients: PolyfitCoefficients
    type: Literal["depth_trend"] = "depth_trend"
    reference_depth: Optional[Array[float]] = None
    max_depth: Optional[Array[float]] = None

    @property
    def effective_depth(self):
        result = np.asarray(self.depth).copy()
        if self.max_depth is not None:
            np.clip(result, None, self.max_depth)
        if self.reference_depth is not None:
            result = result - self.reference_depth
        return result

    def adjust_material(
        self,
        material,
        pressure=None,
        porosity=None,
        mineral=None,
    ):
        """
        :returns: The material at adjusted depth.
        """
        return self.coefficients.polyfit(
            self.effective_depth,
            material,
        )
