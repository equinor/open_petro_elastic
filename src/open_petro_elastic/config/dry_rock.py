import warnings
from dataclasses import field
from typing import List, Union

import numpy as np

try:
    from pydantic.v1 import root_validator
    from pydantic.v1.dataclasses import dataclass
except ImportError:
    from pydantic import root_validator
    from pydantic.dataclasses import dataclass

from .depth_trend import DepthTrend
from .model import Polyfit2dModel, SandstoneModel
from .pressure_dependency import PressureDependency
from .pydantic_config import PetroElasticConfig
from .vector_type import Array

DEFAULT_BULK_COEFFICIENTS = np.array([[2900.0, -1300.0]])
DEFAULT_SHEAR_COEFFICIENTS = np.array([[1700.0, -800.0]])
DEFAULT_PRIMARY_COEFFICIENTS = np.array([[20, -40.0]])
DEFAULT_SECONDARY_COEFFICIENTS = np.array([[12.0, -20.0]])

DEFAULT_DENSITY_COEFFICIENTS = np.array([[0.0, 0.0], [1.0, -1.0]])

DEFAULT_COEFFICIENTS = {
    "density": DEFAULT_DENSITY_COEFFICIENTS,
    "bulk_modulus": DEFAULT_BULK_COEFFICIENTS,
    "shear_modulus": DEFAULT_SHEAR_COEFFICIENTS,
}


@dataclass(config=PetroElasticConfig)
class DryRock:
    """
    Dry rock is a porous material, where the pore framework is made of some
    dense material.
    :param mineral: The mineral in the dense material.
    :param porosity: The fraction of pores to dense material.
    :param model: A SandstoneModel.
    :param coefficients: The coefficients p used for polyfit model. Applied so
        that e.g. bulk_modulus of the dry rock becomes p[i,j] * porosity**i *
        mineral.bulk_modulus ** j.  see MODULI_DEFAULT_COEFFICIENTS and
        VELOCITY_DEFAULT_COEFFICIENTS for defaults.
    """

    model: SandstoneModel = field(
        default_factory=lambda: Polyfit2dModel(DEFAULT_COEFFICIENTS)
    )
    porosity: Array[float] = 0.3

    adjustments: List[Union[DepthTrend, PressureDependency]] = field(
        default_factory=list
    )

    @root_validator
    def add_default_polyfit_coefficents(cls, values):
        if "model" in values and values["model"].type == "polyfit":
            if values["model"].coefficients is None:
                values["model"].coefficients = DEFAULT_COEFFICIENTS
        return values

    @root_validator
    def warn_if_duplicate_adjustment(cls, values):
        num_pressure = 0
        num_depth = 0
        if "adjustments" in values:
            for adj in values["adjustments"]:
                if isinstance(adj, PressureDependency):
                    num_pressure += 1
                if isinstance(adj, DepthTrend):
                    num_depth += 1
        if num_pressure > 1:
            warnings.warn("Adjustments contain more than one pressure dependency")
        if num_depth > 1:
            warnings.warn("Adjustments contain more than one depth trend")
        return values

    @root_validator
    def porosity_does_not_exceed_bound_patchy(cls, values):
        epsilon = 1e-5
        if "model" in values and values["model"].type == "patchy_cement":
            cem_up_bound = (
                values["model"].critical_porosity
                - values["model"].upper_bound_cement_fraction
            )
            if "porosity" in values and np.any(
                values["porosity"] - epsilon > cem_up_bound
            ):
                raise ValueError(
                    "Porosity exceeds critical porosity minus upper"
                    + f"bound of cement fraction: {values['porosity']} vs {cem_up_bound}"
                )
        return values

    @root_validator
    def porosity_does_not_exceed_critical(cls, values):
        epsilon = 1e-5
        if (
            "porosity" in values
            and "model" in values
            and hasattr(values["model"], "critical_porosity")
            and np.any(values["porosity"] + epsilon > values["model"].critical_porosity)
        ):
            raise ValueError(
                f"Porosity exceeds critical porosity: {values['porosity']}"
            )
        return values

    def __post_init_post_parse__(self):
        if hasattr(self.model, "coefficients"):
            if self.model.coefficients.density is None:
                self.model.coefficients.density = DEFAULT_DENSITY_COEFFICIENTS
            if self.model.coefficients.use_moduli:
                if self.model.coefficients.bulk_modulus is None:
                    self.model.coefficients.bulk_modulus = DEFAULT_BULK_COEFFICIENTS
                if self.model.coefficients.shear_modulus is None:
                    self.model.coefficients.shear_modulus = DEFAULT_SHEAR_COEFFICIENTS
            else:
                if self.model.coefficients.primary_velocity is None:
                    self.model.coefficients.primary_velocity = (
                        DEFAULT_PRIMARY_COEFFICIENTS
                    )
                if self.model.coefficients.secondary_velocity is None:
                    self.model.coefficients.secondary_velocity = (
                        DEFAULT_SECONDARY_COEFFICIENTS
                    )

    def nonporous(self, mineral):
        """
        :return: The material representing the dense framework in the
            porous dry rock.
        """
        return self.model.nonporous(mineral)

    def adjust_material(self, material, pressure, mineral):
        for a in self.adjustments:
            material = a.adjust_material(material, pressure, self.porosity, mineral)
        return material

    def material(self, mineral, pressure):
        model_material = self.material_from_model(pressure.truncated_rock, mineral)
        return self.adjust_material(model_material, pressure, mineral)

    def material_from_model(self, effective_pressure, mineral):
        """
        :return: The material representing the dry rock.
        """
        return self.model.eval_material(effective_pressure, self.porosity, mineral)
