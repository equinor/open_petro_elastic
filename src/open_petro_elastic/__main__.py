import argparse
import sys
import warnings
from functools import wraps
from os import path
from shutil import copytree
from traceback import print_tb

import numpy as np
import pandas as pd
import yaml
from pydantic import parse_obj_as

import open_petro_elastic

from .config.input import ColumnInsertionError, Input
from .material import fluid_substitution


def warn_without_traceback(message, category, filename, lineno, file=None, line=None):

    log = file if hasattr(file, "write") else sys.stderr
    log.write(f"Warning: {message}\n")


def calibration_error(calculated, calibrated):
    return (calculated - calibrated) / calibrated


def print_calibration_error(name, error):
    print(
        f"""{name}:
min: {np.min(error)}
mean: {np.mean(error)}
max: {np.max(error)}
"""
    )


def calculate_results(inp):
    mixed_mineral = inp.minerals.as_mixture
    mixed_fluid = inp.fluids.as_mixture(inp.pressure)
    dry_rock_material = inp.dry_rock.material(mixed_mineral, inp.pressure)
    dense_packing = inp.dry_rock.nonporous(mixed_mineral)
    saturated_rock = fluid_substitution(
        dry_rock_material,
        dense_packing,
        mixed_fluid,
        inp.dry_rock.porosity,
    )
    index = [0]
    if inp.data is not None:
        index = inp.data.index
    return pd.DataFrame(
        {
            "ksat": saturated_rock.bulk_modulus,
            "kmin": mixed_mineral.bulk_modulus,
            "kdry": dry_rock_material.bulk_modulus,
            "mysat": saturated_rock.shear_modulus,
            "rsat": saturated_rock.density,
            "kmin_fls": dense_packing.bulk_modulus,
        },
        index=index,
    )


def takes_stream(i, mode):
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            if len(args) > i and args[i] is not None and isinstance(args[i], str):
                with open(args[i], mode) as f:
                    return func(*args[:i], f, *args[i + 1 :], **kwargs)
            else:
                return func(*args, **kwargs)

        return wrapper

    return decorator


class InputError(ValueError):
    pass


def load_config_from_yaml(config_file):
    try:
        config = yaml.safe_load(config_file)
    except yaml.YAMLError as exc:
        if hasattr(exc, "problem_mark"):
            m = "Error while parsing the config file:\n"
            if exc.context is not None:
                m += f"{exc.problem_mark}\n{exc.problem}\n{exc.context}"
            else:
                m += f"\n{exc.problem_mark}\n{exc.problem}"
        else:
            m = "Something went wrong while parsing the config file."
        raise InputError(m) from exc
    if not isinstance(config, dict):
        raise InputError("Unexpected format of config file")
    return config


def load_data_from_csv(data_file):
    try:
        if data_file is not None:
            data = pd.read_csv(data_file)
        else:
            data = None
    except pd.errors.ParserError as e:
        raise InputError("Error while reading data file: {e}") from e
    except pd.errors.EmptyDataError as e:
        raise InputError("Data file seems to be empty") from e
    return data


@takes_stream(0, "r")
def make_input(config_file, data_file):
    """
    :param config_file: Either path to or opened file to yaml config.
    :param data_file: Either path to or opened file to csv data file.
    :return: Input read from config and data
    """
    config = load_config_from_yaml(config_file)
    data = load_data_from_csv(data_file)
    try:
        config["data"] = data
        return parse_obj_as(Input, config)
    except ColumnInsertionError as e:
        raise InputError(f"Error while inserting column data from csv: {e}") from e
    except ValueError as e:
        raise InputError(f"Inconsistent values in input file: {e}") from e


def make_per_row_input(config_file, data_file):
    config = load_config_from_yaml(config_file)
    data = load_data_from_csv(data_file)
    inputs = []
    for i, row in data.iterrows():
        try:
            config["data"] = row.to_frame().transpose()
            inputs.append(parse_obj_as(Input, config))
        except ColumnInsertionError as e:
            print(f"Error while inserting column data from csv: {e}")
            inputs.append(None)
        except ValueError as e:
            print(f"Inconsistent values in input file for line {i}: {e}")
            inputs.append(None)
        except Exception as e:
            print(f"Encountered error while creating input for line {i}: {e}")
            inputs.append(None)
    return inputs


class CalibrationError(ValueError):
    pass


def run_calibration(calibration_file, results):
    try:
        calibration_data = pd.read_csv(calibration_file)
        for key in ["kmin", "kdry", "ksat", "kmin_fls"]:
            result = results[key]
            calibration = calibration_data[key]
            error = calibration_error(result, calibration)
            print_calibration_error(key, error)
    except pd.errors.ParserError as e:
        raise CalibrationError(f"Could not read calibration file: {e}") from e
    except pd.errors.EmptyDataError as e:
        raise CalibrationError("Calibration file seems to be empty") from e
    except (KeyError, IndexError) as e:
        raise CalibrationError(
            f"Error while getting column from calibration data: {e}"
        ) from e


def run_with_threshold(args):
    if args.data_file is None:
        print("In order to use success threshold, data must be given in csv file.")
        return -5
    inputs = make_per_row_input(args.config_file, args.data_file)
    num_failures = 0
    results = pd.DataFrame()
    for idx, inp in enumerate(inputs):
        if inp is None:
            num_failures += 1
            continue
        try:
            results = results.append(calculate_results(inp))
        except Exception as e:
            print(f"Encountered error while calculating rows {id}: {e}")
            num_failures += 1
    if len(inputs) == 0:
        success_percentage = 100
    else:
        success_percentage = (1 - num_failures / len(inputs)) * 100
    if success_percentage < args.success_threshold:
        print(
            f"Percentage of successes {success_percentage} did not reach threshold {args.success_threshold}"
        )
        return -5
    try:
        results.to_csv(args.output_file)
    except Exception as e:
        print(f"Encountered unexpected error while writing output: {e}")
        print_tb(e.__traceback__)
        return -1

    if args.calibration_data is not None:
        try:
            run_calibration(args.calibration_data, results)
        except CalibrationError as e:
            print(f"Error encountered in calibration: {e}")
            return -4
        except Exception as e:
            print(f"Encountered unexpected error during calibration: {e}")
            print_tb(e.__traceback__)
            return -1
    return 0


def run(args):
    if args.success_threshold is not None:
        return run_with_threshold(args)
    try:
        inp = make_input(args.config_file, args.data_file)
    except InputError as e:
        print(f"Could not read input: {e}")
        return -2
    except Exception as e:
        print(f"Encountered unexpected error while reading input: {e}")
        print_tb(e.__traceback__)
        return -1

    try:
        results = calculate_results(inp)
    except ValueError as e:
        print(f"Encountered invalid value in petro elastic calculation: {e}")
        return -3
    except Exception as e:
        print(f"Encountered unexpected error in petro elastic calculation: {e}")
        print_tb(e.__traceback__)
        return -1

    try:
        results.to_csv(args.output_file)
    except Exception as e:
        print(f"Encountered unexpected error while writing output: {e}")
        print_tb(e.__traceback__)
        return -1

    if args.calibration_data is not None:
        try:
            run_calibration(args.calibration_data, results)
        except CalibrationError as e:
            print(f"Error encountered in calibration: {e}")
            return -4
        except Exception as e:
            print(f"Encountered unexpected error during calibration: {e}")
            print_tb(e.__traceback__)
            return -1
    return 0


def generate_tutorial(out_dir):
    here = path.dirname(__file__)
    tutorial_path = path.join(here, "tutorial_config")
    copytree(tutorial_path, out_dir)


class GenerateTutorialAction(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        try:
            generate_tutorial(values)
        except Exception as err:
            print(f"Could not write tutorial config: {err}")
            parser.exit(-4)
        parser.exit()


def percentage(value):
    try:
        value = float(value)
    except (ValueError, TypeError):
        raise argparse.ArgumentTypeError(f"Value must be float {value}")
    if 0.0 <= value <= 100.0:
        return value
    raise argparse.ArgumentTypeError(f"Value must be in range [0, 100] {value}")


TOOL_DESCRIPTION = """
The Equinor petro-elastic modelling tool. Models elastic properties
of sandstone saturated with oil, gas, and brine. See github.com/equinor/open_petro_elastic
for documentation.
"""


def make_parser(prog="open_petro_elastic"):
    ap = argparse.ArgumentParser(prog=prog, description=TOOL_DESCRIPTION)
    ap.add_argument(
        "config_file", type=argparse.FileType("r"), help="petro elastic yaml input file"
    )
    ap.add_argument(
        "--data-file",
        type=argparse.FileType("r"),
        dest="data_file",
        help="Optional additional csv file for data to be inserted into config.",
    )
    ap.add_argument(
        "--calibration",
        type=argparse.FileType("r"),
        dest="calibration_data",
        help="Optional calibration data, if given, an error statistic is calculated and printed.",
    )
    ap.add_argument(
        "--output_file",
        type=argparse.FileType("w"),
        default=sys.stdout,
        help="Optional output file, defaults to stdout.",
    )
    ap.add_argument(
        "--version", action="version", version=open_petro_elastic.__version__
    )
    ap.add_argument(
        "--generate-tutorial",
        action=GenerateTutorialAction,
        help="Write tutorial config. Takes directory to write to as argument.",
    )
    ap.add_argument(
        "--success_threshold",
        type=percentage,
        default=None,
        help="Optional percentage required for success."
        + " Default is 100 percent with no partial results given."
        + " Unsuccessful rows will be given NaN. Note that "
        + "this option can deteriorate performance.",
    )
    return ap


def parse_arguments(argv):
    ap = make_parser(argv[0])
    return ap.parse_args(argv[1:])


def main():
    warnings.showwarning = warn_without_traceback
    sys.exit(run(parse_arguments(sys.argv)))


if __name__ == "__main__":
    main()
