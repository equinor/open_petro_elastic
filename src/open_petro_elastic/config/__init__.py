"""
The open_petro_elastic.config module contains the internal representation of
the configuration file. The parsing of the configuration file goes through
three representations 1) YAML, 2) dictionary of yaml layout 3) Internal representation
as objects built from each section of the config file.

Each of the objects can be built from dictionaries, but it is possible to
construct directly. Generally, the constructor takes the unpacked
corresponding section of the dictionary, ie.  Fluids(**config["fluids"]) where
config is the dictionary from step 2.

"""

from .coefficients import Coefficients
from .constituent import Constituent
from .depth_trend import DepthTrend
from .dry_rock import DryRock
from .fluids import Fluids
from .minerals import Minerals
from .pressure import Pressure
from .pressure_dependency import PressureDependency

__all__ = [
    "DryRock",
    "Pressure",
    "Coefficients",
    "PressureDependency",
    "DepthTrend",
    "Minerals",
    "Constituent",
    "Fluids",
]
