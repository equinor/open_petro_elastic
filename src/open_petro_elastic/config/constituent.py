from typing import Optional

import numpy as np
from typing_extensions import Literal

try:
    from pydantic.v1 import validator
    from pydantic.v1.dataclasses import dataclass
except ImportError:
    from pydantic import validator
    from pydantic.dataclasses import dataclass

from open_petro_elastic.config.fluid_model_providers import fluid_model_providers
from open_petro_elastic.material import Material

from .material_type import MaterialType
from .pydantic_config import PetroElasticConfig
from .vector_type import Array


@dataclass(config=PetroElasticConfig)
class Constituent:
    """
    A Constituent is a part of a composite material to a
    volume or mass fraction.
    :param material: The constituent material which is part of a composite material.
    :param fraction: The ratio of constituent material to composite material.
    """

    material: MaterialType
    fraction: Optional[Array[float]] = None
    fluid_model: Optional[Literal[tuple(fluid_model_providers.keys())]] = None

    def as_material(
        self,
        temperature,
        pressure,
        fluid_model_provider=fluid_model_providers["default"],
    ):
        fluid_model_provider = (
            fluid_model_provider
            if self.fluid_model is None
            else fluid_model_providers[self.fluid_model]
        )
        return self.material.as_material(
            temperature, pressure, fluid_model_provider=fluid_model_provider
        )

    @validator("fraction")
    def fraction_is_ratio(cls, v):
        if v is None:
            return v
        if np.any(v < -1e-5):
            raise ValueError("Constituent fraction must be greater than zero")
        if np.any(v > 1 + 1e-5):
            raise ValueError("Constituent fraction must be less than one")
        return v

    @validator("material", pre=True)
    def wrap_a_material(cls, v):
        if isinstance(v, Material):
            return dict(
                density=v.density,
                bulk_modulus=v.bulk_modulus,
                shear_modulus=v.shear_modulus,
            )
        return v


def fix_one_fraction(constituents):
    """
    Given a list of all constituents of a composite material,
    one of which has fraction set to None, sets the value of
    that missing fraction so that the sum of fractions is 1.0.
    """
    are_none_fraction = [c for c in constituents if c.fraction is None]
    sum_rest = sum([c.fraction for c in constituents if c.fraction is not None])
    epsilon = 1e-5
    if np.any(sum_rest > 1.0 + epsilon):
        raise ValueError(
            f"Sum of constituent fractions {[c.fraction for c in constituents]} should not exceed 1.0, but is {sum_rest}"
        )
    if len(are_none_fraction) == 0:
        return
    if len(are_none_fraction) > 1:
        raise ValueError(
            f"Number of undefined fractions {[c.fraction for c in constituents]} should not be more than 1"
        )
    are_none_fraction[0].fraction = 1.0 - sum_rest
