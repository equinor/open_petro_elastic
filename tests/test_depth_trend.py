from open_petro_elastic.config import DepthTrend, Pressure


def test_dryrock_depthtrend_vp_vs(
    snapshot,
    max_depth,
    depth,
    reference_depth,
    mineral_mix_material,
    modulifit_dryrock,
    depth_coefficients,
):
    modulifit_dryrock.adjustments = [
        DepthTrend(
            max_depth=max_depth,
            depth=depth,
            reference_depth=reference_depth,
            coefficients=depth_coefficients,
        )
    ]

    snapshot.assert_match(modulifit_dryrock.material(mineral_mix_material, Pressure()))


def test_dryrock_depthtrend_k_my(
    snapshot,
    mineral_mix_material,
    max_depth,
    depth,
    reference_depth,
    modulifit_dryrock,
    depth_coefficients,
):
    modulifit_dryrock.adjustments = [
        DepthTrend(
            depth=depth,
            max_depth=max_depth,
            reference_depth=reference_depth,
            coefficients=depth_coefficients,
        )
    ]

    snapshot.assert_match(modulifit_dryrock.material(mineral_mix_material, Pressure()))
