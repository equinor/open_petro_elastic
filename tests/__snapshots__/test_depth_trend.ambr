# serializer version: 1
# name: test_dryrock_depthtrend_k_my
  Material(shear_modulus=[12.], bulk_modulus=[20.], primary_velocity=[0.1165543], secondary_velocity=[0.06729266], density=[2650.])
# ---
# name: test_dryrock_depthtrend_vp_vs
  Material(shear_modulus=[12.], bulk_modulus=[20.], primary_velocity=[0.1165543], secondary_velocity=[0.06729266], density=[2650.])
# ---
